register_buildpack_test.go

```diff
@@ -0,0 +1,130 @@
+package pack
+
+import (
+	"bytes"
+	"context"
+	"os"
+	"os/exec"
+	"runtime"
+	"testing"
+
+	"github.com/buildpacks/imgutil/fakes"
+
+	ifakes "github.com/buildpacks/pack/internal/fakes"
+	"github.com/buildpacks/pack/internal/logging"
+	h "github.com/buildpacks/pack/testhelpers"
+
+	"github.com/heroku/color"
+	"github.com/sclevine/spec"
+	"github.com/sclevine/spec/report"
+)
+
+func TestRegisterBuildpack(t *testing.T) {
+	color.Disable(true)
+	defer color.Disable(false)
+	spec.Run(t, "rebase_factory", testRegisterBuildpack, spec.Parallel(), spec.Report(report.Terminal{}))
+}
+
+func testRegisterBuildpack(t *testing.T, when spec.G, it spec.S) {
+	when("#RegisterBuildpack", func() {
+		var (
+			fakeImageFetcher *ifakes.FakeImageFetcher
+			fakeAppImage     *fakes.Image
+			subject          *Client
+			out              bytes.Buffer
+		)
+
+		it.Before(func() {
+			fakeImageFetcher = ifakes.NewFakeImageFetcher()
+			fakeAppImage = fakes.NewImage("buildpack/image", "", &fakeIdentifier{name: "buildpack-image"})
+
+			h.AssertNil(t, fakeAppImage.SetLabel("io.buildpacks.buildpackage.metadata",
+				`{"id":"heroku/java-function","version":"1.1.1","stacks":[{"id":"heroku-18"},{"id":"io.buildpacks.stacks.bionic"},{"id":"org.cloudfoundry.stacks.cflinuxfs3"}]}`))
+			fakeImageFetcher.RemoteImages["buildpack/image"] = fakeAppImage
+
+			fakeLogger := logging.NewLogWithWriters(&out, &out)
+			subject = &Client{
+				logger:       fakeLogger,
+				imageFetcher: fakeImageFetcher,
+			}
+		})
+
+		it.After(func() {
+			_ = fakeAppImage.Cleanup()
+		})
+
+		it("should return error for an invalid image", func() {
+			fakeAppImage = fakes.NewImage("invalid/image", "", &fakeIdentifier{name: "buildpack-image"})
+			h.AssertNil(t, fakeAppImage.SetLabel("io.buildpacks.buildpackage.metadata", `{}`))
+
+			h.AssertNotNil(t, subject.RegisterBuildpack(context.TODO(),
+				RegisterBuildpackOptions{
+					ImageName: "invalid/image",
+					Type:      "github",
+					URL:       "https://github.com/jkutner/buildpack-registry",
+				}))
+		})
+
+		it("should return error for missing image label", func() {
+			fakeAppImage = fakes.NewImage("missinglabel/image", "", &fakeIdentifier{name: "buildpack-image"})
+			h.AssertNil(t, fakeAppImage.SetLabel("io.buildpacks.buildpackage.metadata", `{}`))
+			fakeImageFetcher.RemoteImages["missinglabel/image"] = fakeAppImage
+
+			h.AssertNotNil(t, subject.RegisterBuildpack(context.TODO(),
+				RegisterBuildpackOptions{
+					ImageName: "missinglabel/image",
+					Type:      "github",
+					URL:       "https://github.com/jkutner/buildpack-registry",
+				}))
+		})
+
+		when("registry type is github", func() {
+			it("should open a github issue in the browser", func() {
+				const expectedURL = `https://github.com/jkutner/buildpack-registry/issues/new?body=%0A%23%23%23+Data%0A%0A%60%60%60toml%0Aid+%3D+%22heroku%2Fjava-function%22%0Aversion+%3D+%221.1.1%22%0Aaddr+%3D+%22buildpack-image%22%0A%60%60%60&title=ADD+heroku%2Fjava-function%401.1.1`
```

Comment:
All references to `jkutner/buildpacks-registry` should be replaces with the `registry.defaultRegistryURL` value.